boards/riscv/tlsr9518adk80d/tlsr9518adk80d.yaml

```diff
@@ -0,0 +1,10 @@
+identifier: tlsr9518adk80d
+name: Telink TLSR9518ADK80D
+type: mcu
+arch: riscv32
+toolchain:
+  - cross-compile
```

Comment:
Can we not use the zephyr toolchain to build this?

Reply:
To build B91 solution Telink specific (NDS32) toolchain is required.
Currently this toolchain is not integrated to SDK-NG project, users need to download this toolchain from Telink website.
The board documentation contains info how to gat the toolchain and how to build.

Reply:
What is unique or different in the telink toolchain?  Is it possible to build a basic version utilizing the common RISC-V support in the Zephyr SDK?  I understand that it may not expose all the features of the hardware, but at least we'd be able to get CI coverage for the platform.

Reply:
> What is unique or different in the telink toolchain?

I do not know what exactly the difference is. I will investigate it and discuss with my colleagues when I come back from vacation in a week, on May 31.

> Is it possible to build a basic version utilizing the common RISC-V support in the Zephyr SDK? I understand that it may not expose all the features of the hardware, but at least we'd be able to get CI coverage for the platform.

No, currently it is not possible. Zephyr SDK contains only RISC-V 64 bits toolchain version, Telink B91 is RISC-V 32.
We already have a task in our backlog to investigate the integration of our toolchain to Zephyr SDK-NG project.
However, we are planning to do this investigation as a second step, the first step is to contribute B91 solution to Zephyr as cross-compile.

Reply:
>No, currently it is not possible. Zephyr SDK contains only RISC-V 64 bits toolchain version, Telink B91 is RISC-V 32.

I am pretty sure we build RV32 platforms with the RV64 toolchain already today, for example [this one](https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/riscv/openisa_rv32m1). Also, #34970 seems to be able to use the Zephyr SDK toolchain to build RV32 images for the GigaDevice SoCs. Could you please find out if there is info we are missing here?

Reply:
I have just checked usage of RISC-V toolchain from zephyr-sdk and the solution is not compilable.
Some registers (which we are using in B91 SoC "start.s" and "soc_irq.s" files) are unknown for this toolchain (mcache_ctl, mmisc_ctl, mxstatus). Also, "hal_telink/tlsr9/drivers/b91/core.h" includes "nds_intrinsic.h" which is part of NDS32 toolchain.

Looks like it is not possible to use the generic RISC-V toolchain since Telink B91 SoC has NDS32 core, which actually is a new architecture based on RISC-V provided by Andes Technology corporation.

@carlescufi, @galak
I propose to release B91 solution as cross-compile first and only after that integrate NDS32 RISC-V toolchain to zephyr-sdk.
Are you OK with that?

Reply:
Wait, but then this is not a RISC-V core, is it? It's actually a new architecture, NDS32? Because I see that [Linux](https://elixir.bootlin.com/linux/latest/source/arch/nds32) has its own arch for it. In this case, we need to add a new arch in `arch/`? CC @stephanosio and @tejlmand

Reply:
OK, I see this in the Andes website: 

>AndesCore processors are based on the AndeStar™ V5 Instruction Set Architecture (ISA), which is compliant to the RISC-V technology.

So I assume that this is, as you mentioned, a customized RISC-V processor. But the NDS32 architecture seems supported by [GCC](https://gcc.gnu.org/backends.html), so I am a bit confused. Adding a requirement for an external toolchain is quite a heavy one, so I'd very much like to get to the bottom of this before we move forward.

Reply:
> Telink B91 SoC has NDS32 core, which actually is a new architecture based on RISC-V

A quick search suggests that "NDS32" refers to the "AndeStar architecture" which was initially based on Andes's own ISA, but later adopted the RISC-V ISA (see http://www.andestech.com/en/products-solutions/andestar-architecture/).

@yurvyn What version of "AndeStar architecture" is TLSR9 based on?

Reply:
[So it seems TLSR9 uses Andes D25F core](https://www.cnx-software.com/2020/11/04/telink-tlsr9-wireless-audio-iot-risc-v-soc-integrates-risc-v-dsp-simd-p-extension/
), which is AndeStar V5 (i.e. RISC-V + Andes extensions).

Reply:
So I've taken a look at the Telink TLSR9 toolchain and it is, without a doubt, binutils/gcc targeting `riscv32-elf`, which is NOT the `nds32-*` target (`nds32-elf` would be for the AndeStar V3 architecture, which is based on Andes's own ISA).

Though, it seems to be heavily patched to support Andes's extension instructions that are not part of the standard RISC-V spec.

See [`nds_intrinsic.h`](https://github.com/andestech/gcc/blob/bsp-v5_1_0/gcc/config/riscv/nds_intrinsic.h).

It looks like these patches have not been upstreamed (or likely there is no plan to do so) and therefore a separate toolchain (possibly `ndsriscv-zephyr-elf`) will be required to support TLSR9.

Reply:
Yes, all of the above is correct.

As I already mentioned, I propose to release B91 solution as cross-compile first and only after that investigate integration of NDS32 RISC-V toolchain to zephyr-sdk. Probably need to contact Andes Technology and ask them to add NDS32 toolchain patches to the upstream.

Reply:
> Probably need to contact Andes Technology and ask them to add NDS32 toolchain patches to the upstream.

That would be ideal, especially noting that many of the intrinsics in `nds_intrinsic.h` seem to be part of the upcoming RISC-V "P extension" which is not finalised yet.

Reply:
>As I already mentioned, I propose to release TLSR9 solution as cross-compile first and only after that investigate integration of NDS32 RISC-V toolchain to zephyr-sdk. Probably need to contact Andes Technology and ask them to add NDS32 toolchain patches to the upstream.

Given that this is currently pending standarization, this sounds like a reasonable compromise, but I'd like the Andes folks, and @galak to chime in as well.

Reply:
@carlescufi Thank you very much for informing this, we haven't found this PR before.

Andes already has Zephyr porting of AndeStar v5 cores (e.g. Andes D25F), and it supports generic RISC-V toolchain if NOT using any V5 extension instructions. This is the [Zephyr source](https://github.com/cwshu/zephyr/tree/andes_v5_demo) of Andes V5 platform (build by `cmake -DBOARD=ae350`)
and entire kernel testcases (`tests/kernel`) can be compiled by generic RISC-V toolchain.

I think TLSR9 SoC could be based on it to support generic RISC-V toolchain (e.g. `soc_v5.h` for custom CSRs). We also plan to upstream Andes V5 platform (soft-cores on the FPGA) to Zephyr community recently.
(p.s. Andes porting is Zephyr v2.4 currently, but I will upgrade SoC/board porting to main branch quickly.)

Andes V5 platform already has 8 Zephyr drivers and has `CONFIG_NOCACHE_MEMORY` support by Andes implementation of RISC-V PMA. Also, we have RISC-V SMP support on Zephyr 2.4 [#29131] and is working on newer version Zephyr.

Reply:
@cwshu,
Thank you very much for the info and for pointing me to the Andes V5 branch.
In this branch I found some useful info about NDS32 arch and now I am able to use the toolchain from zephyr-sdk.

@carlescufi,
I need a few days to finalize the migration to zepkyr-sdk toolchain usage and to finish with the testing.
Will provide new commit soon.

Reply:
Hi @carlescufi

I am testing the new solution (with zephyr-sdk toolchain) and it seems OK, Net Echo Server and Client samples with OpenThread configuration work stable.

However, these is one feature which requires NDS32 toolchain and is not working with zephyr-sdk toolchain. It is CONFIG_FPU (since NDS32 has its own implementation of FPU).

So, we have 2 options:
**Option 1.** We go with NDS32 toolchain as cross-compile and we will ask Andes Technology to add NDS32 toolchain patches to the upstream. 
**Option 2.** We go with zephyr-sdk toolchain and CONFIG_FPU is not supported. In this case we can update the board documentation and add info that to use FPU a custom NDS32 toolchain is required and add the downloadable link to the toolchain.

Cloud you please confirm which option is better for us?

Reply:
@yurvyn Why not both?

Reply:
@yurvyn Could you describe more detail about `CONFIG_FPU` not working with zephyr-sdk? It looks strange.
D25F supports standard RISC-V F/D extension, so that generic toolchain for RVF/RVD can compile and run programs on D25F. For example, I have compiled and run Zephyr FPU test (`tests/kernel/fpu_sharing`) successfully by Zephyr SDK on D25F on the  FPGA.

Reply:
@stephanosio,
> Why not both?

The **Option 2** means support of both toolchains. However, CONFIG_FPU is not supported by the toolchain located in zephyr-sdk. For this option, I propose to update the board documentation with the info that in order to use FPU a custom NDS32 toolchain is required. With the **Option 1** CONFIG_FPU is supported by default.

@cwshu,
> Could you describe more detail about `CONFIG_FPU` not working with zephyr-sdk?

I am using three different toolchains: **NDS32 toolchains** (mculib-v5f and mculib-v5 from AndeSight STD v3.2.3) and the generic RISC-V toolchain which is located in zephyr-sdk (**zephyr-sdk toolchain**). I observe the next behaver:
1. **NDS32 Toolchain (mculib-v5f):** I can successfully build Zephyr TLSR9 solution only in case if `CONFIG_FPU=y`. The test `kernel/fpu_sharing` is pass. In case if `CONFIG_FPU=n` it fails to build even simple samples (hello_world, blinky etc..), there is a linker errors: `can't link single-float modules with soft-float modules`.
2. **NDS32 Toolchain  (mculib-v5):** I can successfully build Zephyr TLSR9 solution only in case if `CONFIG_FPU=n`. In case if `CONFIG_FPU=y` it fails to build even simple samples (hello_world, blinky etc..), there is a linker errors: `can't link single-float modules with soft-float modules`.
3. **zephyr-sdk toolchain**: I can build the Zephyr TLSR9 solution (hello_world, blinky etc..) only in case if `CONFIG_FPU=n`. In case if FPU is enabled, there is a linker errors: `can't link soft-float modules with single-float modules`.

@carlescufi, @stephanosio, @galak
So, looks like the right workflow is: I will remove all NDS32 specific instructions and replace them by the generic RISC-V instructions. As result Zephyr TLSR9 solution could be used with the risc-v toolchain from zephyr-sdk. I will also update the board documentation with the info that in order to use FPU a custom specific NDS32 toolchain is required.

Reply:
Done.
The latest commit supports both zephyr-sdk and cross-compile toolchains.